### PR TITLE
NumpyDataset Tutorial

### PR DESCRIPTION
**Pull request by pvskand**
This is the initial implementation of the tutorial, I will be adding some more concepts and explanations in detail. 
The following are the things to to be present in the tutorial.
- [x] extract `X, y, z` from `NumpyDatatset`
- [x] wrap `X, y, z` from `Numpy` arrays to `NumpyDataset`
- [x] Use `numpy.random` to generate the sample data.
- [x] MNIST dataset as an examples.
- [x] Iterators example
- [x] convert between tf.data and Dataset

### CODE REVIEW

**Comment by rbharath** [comment]
@pvskand Awesome to see this proceeding! As a quick early comment, I'd suggest using smaller matrices (say 4x4) for readability purposes.

**Comment by pvskand** [comment]
@rbharath, sure I will make the matrices smaller. Thanks for the comment. :)

**Comment by lilleswing** [comment]
https://github.com/pvskand/deepchem/blob/master/examples/notebooks/Deepchem_NumpyDataset_tutorial.ipynb

For Readability

**Comment by rbharath** [comment]
@pvskand Took a look at the latest version. This is starting to look nice! Great work :)

**Comment by pvskand** [comment]
@rbharath , @lilleswing , Please have a look at the updated version. I have doubt in the last portion of the tutorial i.e `Converting tf.data to NumpyDataset`, in which I first make a `tf.data` object with random data and labels, and then extracted the numpy arrays form the `tf.data`object via `iterator` and then now since I have the numpy arrays I can easily convert them to `NumpyDataset` object. Is that fine?

File: examples/notebooks/Deepchem_NumpyDataset_tutorial.ipynb [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+    "dataset.w # printing the weights that are assigned by default. Notice that they are a vector of 1's"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 43,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "w = np.random.random((4,)) # initializing weights with random vector of size 20x1\n",
```
Comment (rbharath) [inline examples/notebooks/Deepchem_NumpyDataset_tutorial.ipynb]: s/size 20x1/size 4/

Comment (peastman) [reply examples/notebooks/Deepchem_NumpyDataset_tutorial.ipynb]: `make_iterator()` already covers nearly all use cases where you would care about converting it to a `tf.data.dataset`.

Comment (rbharath) [reply examples/notebooks/Deepchem_NumpyDataset_tutorial.ipynb]: Ah, I see. In that case, adding a bit of documentation on using `make_iterator` would add lots of value here.

Comment (pvskand) [reply examples/notebooks/Deepchem_NumpyDataset_tutorial.ipynb]: @rbharath okay will add that in the tutorial.

Comment (peastman) [reply examples/notebooks/Deepchem_NumpyDataset_tutorial.ipynb]: Also, once I finish up #1106 I'll write a tutorial on the whole Dataset/tf.data/Model/tf.estimator interoperability picture.

**Comment by pvskand** [comment]
@rbharath, @lilleswing Please have a look at the latest version. Added the conversion of `NumpyDataset` to `tf.data` by using `make_iterators()`.

**Comment by rbharath** [comment]
@pvskand Looks good overall. I think this should be ready to merge soon if there aren't further changes you'd like to make. Will wait a bit to see if anyone else has comments as well.

**Comment by rbharath** [comment]
@pvskand Ok to merge in this PR? Could you remove the "[WIP]" tag if so

**Comment by pvskand** [comment]
@rbharath Done.

**Comment by rbharath** [comment]
@pvskand Congrats on an impressive contribution! I think this tutorial will add a lot of value to our users and will prove a great reference.

LGTM

**Comment by pvskand** [comment]
@rbharath, the link to notebook doesn't appear https://deepchem.io/docs/notebooks/index.html. Can you add the link on the webpage.

**Comment by rbharath** [comment]
@pvskand Could you make a pull-request to the https://github.com/deepchem/deepchem.io website repo? You will need to modify https://github.com/deepchem/deepchem.io/blob/master/website/docs/notebooks/index.html to add a link to the new tutorial. This will automatically be built and deployed once merged.

**Comment by pvskand** [comment]
@rbharath, Sure.

**Pull request merged**